src/Symfony/Component/EventDispatcher/Debug/TraceableEventDispatcher.php

```diff
@@ -137,28 +147,46 @@ public function dispatch($eventName, Event $event = null)
         $this->preDispatch($eventName, $event);
 
         $e = $this->stopwatch->start($eventName, 'section');
+        $previousRequestHash = $this->currentRequestHash;
 
-        $this->dispatcher->dispatch($eventName, $event);
+        try {
+            $this->dispatcher->dispatch($eventName, $event);
+            $this->currentRequestHash = $previousRequestHash;
 
-        if ($e->isStarted()) {
-            $e->stop();
-        }
+            if ($e->isStarted()) {
+                $e->stop();
+            }
+
+            $this->postDispatch($eventName, $event);
+        } finally {
+            $this->currentRequestHash = $previousRequestHash;
 
-        $this->postDispatch($eventName, $event);
-        $this->postProcess($eventName);
+            if ($e->isStarted()) {
+                $e->stop();
+            }
+
+            $this->postProcess($eventName);
+        }
 
         return $event;
     }
 
     /**
      * {@inheritdoc}
      */
-    public function getCalledListeners()
+    public function getCalledListeners(/* Request $request = null */)
     {
+        if (null === $this->callStack) {
+            return array();
+        }
+
+
+        $hash = 1 <= \func_num_args() && null !== ($request = \func_get_arg(0)) ? spl_object_hash($request) : null;
         $called = array();
-        foreach ($this->called as $eventName => $listeners) {
-            foreach ($listeners as $listener) {
-                $called[$eventName.'.'.$listener->getPretty()] = $listener->getInfo($eventName);
+        foreach ($this->callStack as $listener) {
+            list($eventName, $requestHash) = $this->callStack->getInfo();
+            if (null === $hash || $hash === $requestHash) {
+                $called[] = $listener->getInfo($eventName);
```

Comment:
they key caused dedupping, which we dont want. So we need to change it anyway (include the hash) but i favored numerical indexes.